### PR TITLE
Breaking change: Disallow raw-string-literal interpolations at start of line when indentation-whitespace is present.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Dependent on the decision from LDM if we will take this break or not.

Previously, the following was erroneously allowed:

```csharp
var x = $"""
    Hello
{1 + 1}
    World
    """;
```

This violated the rule that the lines content (including where an interpolation starts) must start with same whitespace as the final `    """;` line.  It is now required that the above be written as:

```csharp
var x = $"""
    Hello
    {1 + 1}
    World
    """;
```

[Sharplab](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcDMACR2DC2A3utmbkgGy4As2AsgIYCWUAFAJTGnm8BujMNj7YAvNgAkAIhk9e87ADMA9srkLyRbEmwBqbdgC+6jdgBGgkxplSA3FdNkHFAJxs+He2l7G0hoA==)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs [hunk 1]
```diff
@@ -243,6 +239,69 @@ kind switch
...
+            //          {1 + 1}
+            //          """
+            //
+            // Not:
+            //
+            //      var x = """
+            // {1 + 1}
+            //          """
+            //
+            // Note: We don't need to check
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs:282]: The example that we don't need to check is exactly the same as the example just above, which we are checking? I'm not sure how I'm supposed to read this comment.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests.cs [hunk 2]
```diff
@@ -314,6 +314,58 @@ static void Main(string[] args)
...
+            // (12,1): error CS8999: Line does not start with the same whitespace as the closing line of the raw string literal.
+            // {
+            Diagnostic(ErrorCode.ERR_LineDoesNotStartWithSameWhitespace, "{").WithLocation(12, 1));
+    }
+
+    [Fact]
+    public void TwoInserts02_A()
+    {
+        string source =
+@"using System;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests.cs:328]: I'm surprised you're not using a raw string for this test too :).

**Review by 333fred — COMMENTED** [review]
Just some comment confusion. LDM appears to be accepting of this change.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs [hunk 3]
```diff
@@ -141,7 +149,11 @@ ReadOnlySpan<char> getIndentationWhitespace(ReadOnlySpan<char> originalTextSpan)
             }
 
             InterpolatedStringContentSyntax? makeContent(
-                ReadOnlySpan<char> indentationWhitespace, StringBuilder content, bool isFirst, bool isLast, ReadOnlySpan<char> text)
+                ReadOnlySpan<char> indentationWhitespace,
+                StringBuilder content,
+                bool isFirst,
+                bool isLast,
+                ReadOnlySpan<char> text)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs:156]:
```suggestion
                ReadOnlySpan<char> indentationWhitespace, StringBuilder content, bool isFirst, bool isLast, ReadOnlySpan<char> text)
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs:268]: the example below has leading whitespace.  i'll doc it to make it clearer.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs:268]: doc'ed

File: src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests.cs:328]: :D.  when in rome.  I would be ok making a larger change to update these later.  Esp. if we may consider taking this backwards so we get the breaking change as soon as possible.

**Comment by CyrusNajmabadi** [comment]
> We might want to document it in the breaking change doc though.

Happy to.  Can you link me to that doc and i can make the change.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RawInterpolatedStringLiteralCompilingTests.cs [hunk 4]
```diff
@@ -1248,6 +1248,40 @@ public void MultiLineCase34()
...
     }
 
+    [Fact]
+    public void MultiLineCase34_A()
+    {
+        RenderAndVerify(@"
+System.Console.Write(
+    $""""""
+    a
+{42}
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RawInterpolatedStringLiteralCompilingTests.cs:1258]: Do we already have a test that shows the expression in the interpolation hole doesn't need to have the same leading whitespace?

```cs
var s = """
    hello {
1 + 2 }
    """;

var s = """
    hello
    {
1 + 2
} world
    """;
```

I assume the `{` that begins the hole needs to be properly indented. What about if the line starts with the `}` that closes a hole? Does that also need to be indented? Maybe not, and it's sort of treated as being part of the same line of content as the `{`?

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/RawInterpolatedStringLiteralCompilingTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RawInterpolatedStringLiteralCompilingTests.cs:1258]: > Do we already have a test that shows the expression in the interpolation hole doesn't need to have the same leading whitespace?

Yes.

> What about if the line starts with the } that closes a hole?

That's an interesting question.  IMO, that whitespace is part of the expression.  It's *inside* the interpolation.  As such, it is not under the same rule.

**Review by RikkiGibson — APPROVED** [review]

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md [hunk 5]
```diff
@@ -1,5 +1,31 @@
...
+```csharp
+var x = $"""
+    Hello
+{1 + 1}
+    World
+    """;
+```
+
+This violated the rule that the lines content (including where an interpolation starts) must start with same whitespace as the final `    """;` line.  It is now required that the above be written as:
+
```
Comment (jcouv) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:18]: nit: double empty line (also applies to line 27)

**Comment by CyrusNajmabadi** [comment]
LDM raised no issues taking this change.  Merging in.

**Comment by jcouv** [comment]
Please remember to squash compiler PRs. Thanks

**Comment by Korporal** [comment]
Does the fix to raw string literals also catch this - currently legal:

image

**Comment by CyrusNajmabadi** [comment]
That's fully legal and intentionally represents the contents:

```
<space><space>I am text<newline>
<space><space>so am I
```

**Comment by Korporal** [comment]
> That's fully legal and intentionally represents the contents:
> 
> ```
> <space><space>I am text<newline>
> <space><space>so am I
> ```

I see, well in that case the diagnostic message is imprecise, if I do this:

```cs
        """
        I am text +
        so am I +
         """;
```

I get: 

`Line does not start with the same whitespace as the closing line of the raw string literal.`

Yet this also does not meet that criteria, yet is deemed legal:

```cs
        """
        I am text +
        so am I +
       """;
```

Should the diagnostic message not be something like:

`Line must not start with shorter whitespace than the closing line of the raw string literal.`

?

**Comment by CyrusNajmabadi** [comment]
> Line does not start with the same whitespace as the closing line of the raw string literal.`

That is correct.  Each line of content needs to start with teh same whitespace as the closing line of the raw string literal.

> Yet this also does not meet that criteria, yet is deemed legal:

It is legal, because each content line starts with teh same whitespace as the closing line of the raw string literal.

> Line must not start with shorter whitespace than the closing line of the raw string literal.

That's what "Starts with" means.  It's the same semantics as [StartsWith](https://learn.microsoft.com/en-us/dotnet/api/system.string.startswith?view=net-7.0#system-string-startswith(system-string))

**Comment by Korporal** [comment]
@CyrusNajmabadi - I understand thanks. The error message seems a little cryptic then. Visually the lines above "start with" 4 spaces and in the two examples the end line "starts with" 5 spaces and 3 spaces respectively, so the "start" whitespace of each line is not "the same". Nothing in the error message refers to the semantics of the `StartsWith` method either.

I suggest:

**"Any sequence of whitespace characters preceding the closing line of the raw string literal must also appear at start of every other line within the string literal."**

The more precisely, the more informative the error message is worded in English too, the more likely it will be translated to another language, without loss of meaning, less scope for misunderstanding.

I also searched the web quickly but could not find any documentation on CS8999 that explains the error.

**Comment by CyrusNajmabadi** [comment]
> Visually the lines above "start with" 4 spaces and in the two examples the end line "starts with" 5 spaces and 3 spaces respectively,

Right.  So '4' spaces 'start[s] with the same whitespace' ('3' spaces) in this legal case.  :)

But '4' spaces' does not 'start with the same whitespace' ('5' spaces) in the illegal case. :)

> othing in the error message refers to the semantics of the StartsWith method either.

I mean.  it uses the words 'start with'.  :)

I pointed out the .net api here because it just shows that htis terminology is bog standard and consistent.  

> The more precisely, the more informative the error message 

I think the terminology is fine. :)

"Line does not start with the same whitespace as the closing line of the raw string literal."

You get the error in the right location, and it states what the issue is.  We even give a more precise error if there are strange deviations in whitespace (for example, using a tab in one line, but normal spaces in the other.  

But thanks for hte feedback!  If we hear from others about confusion, we may consider changing things.  But so far this has not appeared to be a problem, so i think keeping the terminology simple, clear and to the point, is good here.

**Comment by Korporal** [comment]
As you wish, but if it _were_ clear I'd never have suspect a bug in the first place.

**Pull request merged**